components/style/values/computed/basic_shape.rs

```diff
@@ -10,10 +10,43 @@
 use cssparser::ToCss;
 use properties::shorthands::serialize_four_sides;
 use std::fmt;
+use url::Url;
 use values::computed::position::Position;
 use values::computed::{BorderRadiusSize, LengthOrPercentage};
+use values::computed::UrlExtraData;
 
-pub use values::specified::basic_shape::FillRule;
+pub use values::specified::basic_shape::{FillRule, GeometryBox, ShapeBox};
+
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum ShapeSource<T> {
+    Url(Url, UrlExtraData),
+    ShapeOrBox(Option<BasicShape>, Option<T>),
```

Comment:
I think it would be clearer to represent shape values and "none" values as separate enum values.  So:

```
  None,
  Url(Url, UrlExtraData),
  Shape(BasicShape, Option<T>),
  Box(T),
```